### ISSUE TITLE
S4 Listops have nan loss

### ISSUE THREAD

**Issue by lthilnklover** [HazyResearch/state-spaces#138]
First of all, thank you for the comprehensive code base for all variants of S4 models.

However, as I try to run the Listops experiments with S4 (HYYT version), the losses for train, test and val all become nan after 1 epoch. 

I ran the following script:

`python -m train experiment=lra/s4-listops wandb=null`

The final accuracy is also way below the reported accuracy (train=0.17).

Is there something that I have done wrong..?

**Comment by agostinv** [comment]
Suffering from a similar issue. Admittedly, I ported the example S4D model (featuring the specified custom Dropout) into the Skyformer LRA repo: https://github.com/pkuzengqi/Skyformer. The model is essentially 1:1 with the one in this repo with an added embedding layer at the beginning and linear classifier at the end (naturally) in addition to some added masking behavior. 

Nonetheless, dev accuracy finishes at around 0.18 or so regardless of my changes.

**Comment by radarFudan** [comment]
I came across the same problem and decreasing learning rate by 10 cannot solve this problem.

**Comment by bngcode** [comment]
Same problem here. I am using a completely different dataset for audio processing. I extracted the S4ND and S4 layers into a different neural network architecture and I also got NaN after one epoch because the self.log_dt in SSKernelNPLR is nan. This must have happened during backpropagation because it is not updated otherwise (I believe)?

**Comment by albertfgu** [comment]
Sorry for not responding to this. I don't know why this is happening. I haven't revisited these experiments in a long time, but I'm quite confident that they were reproducible in the past. Perhaps something has changed in the libraries or perhaps there are some numerical issues on certain hardware

**Comment by wwy155** [comment]
same problem, here is a solution to circumvent by changing the SSKernelNPLR class:

```python
with torch.no_grad():
    # Increase the internal length if needed
    while rate * L > self.L:
        self.double_length()
    dt = torch.exp(self.log_dt) * rate
    B = _r2c(self.B)
    C = _r2c(self.C)
    P = _r2c(self.P)
    Q = P.conj() if self.Q is None else _r2c(self.Q)
    w = self._w()
```
I don't know whether this will be detrimental to the performance or not, at least, no nan ever reported.

**Comment by chreissel** [comment]
I have the same problem. @icannotnamemyself, could you comment a bit more on your solution? I am not sure I understand exactly the reasoning or where to make the modifications.
